### PR TITLE
Make sample extra policy work with unknown policies

### PR DESCRIPTION
**Pull request by henschkowski**
The sample extra check plugin fails if the checked apiproxy has an unknown policyType in it, like "DecodeJWT". This change makes the plugin ignore those policies.

### CODE REVIEW

**Comment by henschkowski** [comment]
added the cla info

**Comment by ssvaidyanathan** [comment]
Isn't that the correct approach - fail if an unknown policy is added? Please add an issue when the plugin needs an update to accommodate a new policy

@DinoChiesa - WDYT?

**Comment by kurtkanaskie** [comment]
I just noticed this. Can you provide an example of the situation you are fixing?

For example, I changed a policy from `<CORS ...` to `<CORS-X ...` and apigeelint didn't complain, in fact it totally ignored the policy.

Is that the issue? What version of apigeelint did you use? I'm using 2.11.0

**Comment by henschkowski** [comment]
Was giving a demo to a customer, showing that they could include their own check plugins with that --externalPluginsDirectory option. But the program just did not pick up that plugin - at least that was the experience. I used a simple proxy with a DecodeJWT policy.

IMHO the plugin should just ignore that unknown policy and move on (better: log it / warn at the end of output).

**Comment by kurtkanaskie** [comment]
Hi henschkowski,
Please try the same on version 2.11.0, which I just tested.

I'm still not following your use case. Running the following in the source directory:
`apigeelint -x ./externalPlugins -e PO007 -s ./apiproxy`
Uses the `EX-PO007-NamingConventions.js` external plugin and excludes the use of the built in plugin `PO007`.

If there is a policy (e.g. CORS) which that external plugin does not know about, it quietly ignores it.

Furthermore, if I use the classic Apigee profile (default) I am warned about a policy that is not available for that profile (e.g. CORS), but that's unrelated.

```
/apiproxy/policies/CORS.xml

║ Line     │ Column   │ Type     │ Message                                                │ Rule ID              ║
╟──────────┼──────────┼──────────┼────────────────────────────────────────────────────────┼──────────────────────╢
║ 0        │ 0        │ warning  │ The policy type (CORS) is not available in the         │ PO028                ║
║          │          │          │ profile apigee.                                        │                      ║


```

**Comment by henschkowski** [comment]
Hello kurtkanaskie,

Tried with 2.11.0 (git master), same result. Maybe it depends on the position of the policy, I test with a DecodeJWT policy right at the beginning.

Warnings from the EX-P0007 plugin will then not show up, as the plugin is quietly ignored **for all policies**. With the change that I proposed those warnings will show up, albeit only for known policies. 

It's not a big thing, just making it more intuitive for first-time users, IMHO.

**Comment by kurtkanaskie** [comment]
OK, I'm trying to reproduce the issue and demonstrate your fix.\
I ran the following on the attached proxy. 
```
./cli -x ./externalPlugins -s ./apiproxy -f table.js --profile apigeex
```
Kindly demonstrate what you've fixed, because I'm still not seeing it
[apiproxy.zip](https://github.com/apigee/apigeelint/files/7664669/apiproxy.zip)

**Comment by henschkowski** [comment]
That apiproxy does not show it as it contains two "unknown" ### policies.
Please add an "Extract-Variables" policy like this and retry:

Extract-Variables-1.xml: 

<pre>&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot; standalone=&quot;yes&quot;?&gt;
&lt;ExtractVariables name=&quot;Extract-Variables-1&quot;&gt;
    &lt;DisplayName&gt;Extract Variables-1&lt;/DisplayName&gt;
    &lt;Properties/&gt;
    &lt;Header name=&quot;Authorization&quot;&gt;
        &lt;Pattern ignoreCase=&quot;false&quot;&gt;Bearer {private.jwt}&lt;/Pattern&gt;
    &lt;/Header&gt;
    &lt;Source clearPayload=&quot;false&quot;&gt;request&lt;/Source&gt;
</pre>



You should see (without the change):
<pre>/apiproxy/policies/CORS.xml

║ <b>Line</b>     │ <b>Column</b>   │ <b>Type</b>     │ <b>Message</b>                                                │ <b>Rule ID</b>              ║
╟──────────┼──────────┼──────────┼────────────────────────────────────────────────────────┼──────────────────────╢
║ 2        │ 51       │ <font color="#C4A000">warning</font>  │ Non-standard prefix (null). Valid prefixes for         │ PO007                ║
║          │          │          │ CORS include: [&quot;cors&quot;]                                 │                      ║
║ 0        │ 0        │ <font color="#C4A000">warning</font>  │ The policy type (CORS) is not available in the         │ PO028                ║
║          │          │          │ profile apigee.                                        │                      ║

/apiproxy/policies/Extract-Variables-1.xml

║ <b>Line</b>     │ <b>Column</b>   │ <b>Type</b>     │ <b>Message</b>                                                │ <b>Rule ID</b>              ║
╟──────────┼──────────┼──────────┼────────────────────────────────────────────────────────┼──────────────────────╢
║ 0        │ 0        │ <font color="#C4A000">warning</font>  │ Extract-Variables-1 is not attached to a Step in       │ BN005                ║
║          │          │          │ the bundle.  Remove unused policies.                   │                      ║
║ 3        │ 5        │ <font color="#C4A000">warning</font>  │ Filename &quot;Extract-Variables-1.xml&quot; does not match      │ PO008                ║
║          │          │          │ policy display name &quot;Extract Variables-1&quot;. To          │                      ║
║          │          │          │ avoid confusion when working online and offline        │                      ║
║          │          │          │ use the same name for files and display name in        │                      ║
║          │          │          │ policies (excluding .xml extension).                   │                      ║

╔════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║ <font color="#CC0000">0 Errors</font>                                                                                                       ║
╟────────────────────────────────────────────────────────────────────────────────────────────────────────────────╢
║ <font color="#C4A000">4 Warnings</font>                                                                                                     ║
╚════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝

</pre>

And with the change:

<pre>/apiproxy/policies/CORS.xml

║ <b>Line</b>     │ <b>Column</b>   │ <b>Type</b>     │ <b>Message</b>                                                │ <b>Rule ID</b>              ║
╟──────────┼──────────┼──────────┼────────────────────────────────────────────────────────┼──────────────────────╢
║ 2        │ 51       │ <font color="#C4A000">warning</font>  │ Non-standard prefix (null). Valid prefixes for         │ PO007                ║
║          │          │          │ CORS include: [&quot;cors&quot;]                                 │                      ║
║ 0        │ 0        │ <font color="#C4A000">warning</font>  │ The policy type (CORS) is not available in the         │ PO028                ║
║          │          │          │ profile apigee.                                        │                      ║

/apiproxy/policies/Extract-Variables-1.xml

║ <b>Line</b>     │ <b>Column</b>   │ <b>Type</b>     │ <b>Message</b>                                                │ <b>Rule ID</b>              ║
╟──────────┼──────────┼──────────┼────────────────────────────────────────────────────────┼──────────────────────╢
║ 0        │ 0        │ <font color="#C4A000">warning</font>  │ Extract-Variables-1 is not attached to a Step in       │ BN005                ║
║          │          │          │ the bundle.  Remove unused policies.                   │                      ║
║ 3        │ 5        │ <font color="#C4A000">warning</font>  │ Filename &quot;Extract-Variables-1.xml&quot; does not match      │ PO008                ║
║          │          │          │ policy display name &quot;Extract Variables-1&quot;. To          │                      ║
║          │          │          │ avoid confusion when working online and offline        │                      ║
║          │          │          │ use the same name for files and display name in        │                      ║
║          │          │          │ policies (excluding .xml extension).                   │                      ║
║ 0        │ 0        │ <font color="#C4A000">warning</font>  │ Naming Conventions: Policy &quot;Extract Variables-1&quot;       │ EX-PO007             ║
║          │          │          │ of type &quot;ExtractVariables&quot; should have an              │                      ║
║          │          │          │ indicative prefix. Valid prefixes include: [&quot;EV-&quot;]     │                      ║

╔════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║ <font color="#CC0000">0 Errors</font>                                                                                                       ║
╟────────────────────────────────────────────────────────────────────────────────────────────────────────────────╢
║ <font color="#C4A000">5 Warnings</font>                                                                                                     ║
╚════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝
</pre>

**Comment by kurtkanaskie** [comment]
Ah, now I see it, thanks, that's a subtle one.

In my words:\
The sample external plugin `EX-PO007-NamingConventions.js` errors and exits quietly if the checked apiproxy has a policyType in it that is not represented in it's `policyMetaData` variable, like "DecodeJWT". This change makes the plugin skip those unmentioned policies.

The error:
```
apigeelint:bundleLinter plugin error: EX-PO007-NamingConventions.js TypeError: Cannot read property 'indications' of undefined +0ms
```

File: externalPlugins/EX-PO007-NamingConventions.js [hunk 1]
```diff
@@ -66,9 +66,17 @@ var plugin = {
...
     found = false,
-    hadWarn = false;
+    hadWarn = false,
+    policyMeta = policyMetaData[policyType];
+    
+  if (!policyMeta) {
+      // No prefix defined for this policy
+      found=false;
+      hadWarn=false;
+      return;
```
Comment (DinoChiesa) [inline externalPlugins/EX-PO007-NamingConventions.js:77]: We must  not return directly from  onPolicy without invoking the callback.  The convention is to check the cb to see if it is a function, then invoke it with the `hadWarn` flag either true or false as appropriate.  Also there's no need to set the booleans to the same values they already have.

File: externalPlugins/EX-PO007-NamingConventions.js [hunk 2]
```diff
@@ -66,9 +66,17 @@ var plugin = {
...
+  }
+  var prefixes = policyMeta.indications;
 
   prefixes.some(function(prefix) {
     if (displayName.startsWith(prefix)) {
```
Comment (DinoChiesa) [inline externalPlugins/EX-PO007-NamingConventions.js:82]: This code is confusing and inelegant.  Should be updated to use modern conventions. Eg
```
found = prefixes.some( p => displayName.startsWith(p)); 
```

File: externalPlugins/EX-PO007-NamingConventions.js [hunk 3]
```diff
@@ -66,9 +66,17 @@ var plugin = {
 var onPolicy = function(policy, cb) {
   var displayName = policy.getDisplayName(),
```
Comment (DinoChiesa) [inline externalPlugins/EX-PO007-NamingConventions.js:67]: This example plugin should not examine the display name, but rather the name.  If we really want the example to check the display name, then the plugin file name should be changed to reflect that purpose.

File: externalPlugins/EX-PO007-NamingConventions.js [hunk 4]
```diff
@@ -66,9 +66,17 @@ var plugin = {
 var onPolicy = function(policy, cb) {
```
Comment (DinoChiesa) [inline externalPlugins/EX-PO007-NamingConventions.js:66]: The problem of the "missing metadata" as noted in PR #273 is valid.  And also, this code has other problems which should also be fixed.

**Review by DinoChiesa — COMMENTED** [review]
We cannot merge the suggested change without other changes.  But I've submitted #284 based on this suggestion, I think we should just merge that.

**Comment by ssvaidyanathan** [comment]
Closing this as #284 has the fix for this

@kurtkanaskie - can you please run the same test with the latest code from master

**Comment by kurtkanaskie** [comment]
Ran tests using latest code and it works.

**Comment by ssvaidyanathan** [comment]
v2.12.0 has this fix

**Pull request closed without merging**